guidelines/CPP_STYLE_GUIDE.md

```diff
@@ -0,0 +1,221 @@
+};
+
+class MyClass : public IMyVirtualClass {
+public:
+  void my_virtual_function() override;
+  float get_private_data() const;
+  float public_data_member;
+protected:
+  float protected_data_member_;
+private:
+  float private_data_member_;
+};
+
+void my_function(const MyClass& readonly_reference, MyClass& mutable_reference) {
+  int my_local_variable;
+}
+```
+
+
+### Format and whitespace
+
+- Indents are 2 spaces.
+- Conditionals and loops always have brackets, with `else if` / `else` statements following the closing bracket on the same line. No spaces inside parentheses.
+- In a `for` loop, place a space after `;`, and on both sides of `:`.
+- `switch` statements should define a default case. Only use additional scoping brackets within a case when necessary (e.g. declaring a local variable).
+
+```c++
+if (condition) {
+  foo();
+} else if {
+  bar();
+} else {
+  ...
+}
+
+while (condition) {
+  foo();
+}
+
+for (int i = 0; i < 10; ++i) {
+  foo();
+}
+
+std::vector<SomeType> range(...);
+for (auto iter : range) {
+  bar();
+}
+
+switch (value) {
+  case 0:
+    foo();
+    break;
+  case 1: {
+    float local_var;
+    bar();
+    break;
+  }
+  default:
+  case 2:
+    foo();
+    bar();
+    break;
+}
+```
+
+Pointer and reference operators `*` and `&` should follow the type. 
+```c++
+int a = 0;
+int* b = &a;
+int& c = a;
+int& d = *b;
+```
+
+Prefer smart pointers to raw pointers whenever possible.
+
+## Automatic formatting
+
+While it is useful to internalize the rules of this style guide while writing your code, it would be cumbersome and unrealistic to check every line against every rule manually. Fortunately, the majority of the formatting rules can be applied to your code automatically. By configuring your development environment to automatically format your code (for example, format on save), you can ensure that each new addition you make will conform to the lab standard without any effort on your part. 
+
+Some files have been provided to help automatically reformat your code according to the AICA style. Use the AICA style directly within your code editor by importing the [IntelliJ IDEA code style XML](https://github.com/epfl-AICA/wiki/blob/master/files/AICA%20IntelliJ%20IDEA%20CPP%20Code%20Style.xml). Alternatively, post-format your code using your editor or the `clang-format` command-line tool with the [clang-format-AICA file](https://github.com/epfl-AICA/wiki/blob/master/files/clang-format-AICA).
+
+Follow the setup instructions below and within less than 5 minutes, your favourite code editor will automatically format your code in the future.
+
+_(Instructions written for Ubuntu 18.04)_
+
+### CLion
+Download the [IntelliJ IDEA code style XML file](./AICA_IDEA_CPP_STYLE.xml) from this repository.
+
+In CLion, go to `File` &rarr; `Settings` &rarr; `Editor` &rarr; `Code Style`. Then, under `Scheme`, click on `Show Scheme Actions` and then `Import Scheme...`. Select the file you downloaded and confirm with `OK`. 
```

Comment:
nitpick: I would just start with `Settings` here and remove file, like that its also good on Mac